src/api/c/memoryapi.hpp

```diff
@@ -0,0 +1,96 @@
+/*******************************************************
+ * Copyright (c) 2019, ArrayFire
+ * All rights reserved.
+ *
+ * This file is distributed under 3-clause BSD license.
+ * The complete license agreement can be obtained at:
+ * http://arrayfire.com/licenses/BSD-3-Clause
+ ********************************************************/
+
+#pragma once
+
+#include <common/MemoryManager.hpp>
+
+#include <events.hpp>
+#include <af/event.h>
+#include <af/memory.h>
+
+////////////////////////////////////////////////////////////////////////////////
+// Buffer Info
+////////////////////////////////////////////////////////////////////////////////
+
+struct BufferInfo {
+    void *ptr;
+    af_event event;
+};
+
+BufferInfo &getBufferInfo(const af_buffer_info pair);
+
+af_buffer_info getHandle(BufferInfo &pairHandle);
+
+detail::Event &getEventFromBufferInfoHandle(const af_buffer_info handle);
+
+////////////////////////////////////////////////////////////////////////////////
+// Memory Manager API
+////////////////////////////////////////////////////////////////////////////////
+
+struct MemoryManager {
+    // Callbacks from public API
+    af_memory_manager_initialize_fn initialize_fn;
+    af_memory_manager_shutdown_fn shutdown_fn;
+    af_memory_manager_alloc_fn alloc_fn;
+    af_memory_manager_allocated_fn allocated_fn;
+    af_memory_manager_unlock_fn unlock_fn;
+    af_memory_manager_garbage_collect_fn garbage_collect_fn;
+    af_memory_manager_print_info_fn print_info_fn;
+    af_memory_manager_usage_info_fn usage_info_fn;
+    af_memory_manager_user_lock_fn user_lock_fn;
+    af_memory_manager_user_unlock_fn user_unlock_fn;
+    af_memory_manager_is_user_locked_fn is_user_locked_fn;
+    af_memory_manager_get_mem_step_size_fn get_mem_step_size_fn;
+    af_memory_manager_get_max_bytes_fn get_max_bytes_fn;
+    af_memory_manager_get_max_buffers_fn get_max_buffers_fn;
+    af_memory_manager_set_mem_step_size_fn set_mem_step_size_fn;
+    af_memory_manager_check_memory_limit check_memory_limit_fn;
+    af_memory_manager_add_memory_management add_memory_management_fn;
+    af_memory_manager_remove_memory_management remove_memory_management_fn;
+    // A generic payload on which data can be stored on the af_memory_manager
+    // and is accessible from the handle
+    void *payload;
+};
+
+MemoryManager &getMemoryManager(const af_memory_manager manager);
+
+af_memory_manager getHandle(MemoryManager &manager);
+
+/**
+ * An internal wrapper around an af_memory_manager which calls function pointers
+ * on a af_memory_manager via calls to a MemoryManagerBase
+ */
+class MemoryManagerFunctionWrapper : public common::memory::MemoryManagerBase {
+    af_memory_manager handle_;
```

Comment:
This probably won't work — if we combine these into a single `CustomMemoryManager` that acts as the handle, since we're passing around a pointer, we need to give each backend's `DeviceManager` a way of holding onto the memory manager that is aware of its lifetime (this is why the current API uses `unique_ptr`)

The way to do that would be to wrap a `CustomMemoryManager` in a `unique_ptr` when we set it as the default memory manager. There are two issues with that:
- The `af_memory_manager` is no longer decoupled from the lifetime of the thing that the `DeviceManager` holds onto - this means that if you set a new memory manager, an `af_memory_manager` you implicitly held onto may be destroyed without your knowing. This makes reconciling maintenance of internal state of a user-defined `af_memory_manager` with when the internal API destroys it difficult.
- The destruction order with each backend's `NativeMemoryInterface` gets messed up because each backend's `NativeMemoryInterface` calls `shutdown` on the `MemoryManagerBase` it belongs to. But if that thing is destroyed (because we're storing it on `DeviceManager` directly), the call to its `shutdown` becomes pure virtual. That can't be fixed unless we decouple the actual `shutdown` from the wrapper shutdown, which the current API accomplishes